core/src/main/java/org/infinispan/manager/DefaultCacheManager.java

```diff
@@ -1139,11 +1138,14 @@ public ClusterExecutor executor() {
       if (transport != null) {
          long time = configurationManager.getGlobalConfiguration().transport().distributedSyncTimeout();
          return ClusterExecutors.allSubmissionExecutor(null, this, transport, time, TimeUnit.MILLISECONDS,
-               globalComponentRegistry.getComponent(ExecutorService.class, KnownComponentNames.REMOTE_COMMAND_EXECUTOR),
+               // This can run arbitrary code, including user - such commands can block
```

Comment:
No, I more put it here because it can block :)

Reply:
Should it be using the non-blocking executor then?

Reply:
Unfortunately until the other JIRA is fixed, we don't have a great solution. And cluster executor isn't that widely used afaik. But we should hopefully get it fixed before that.

Reply:
Just an FYI but the only way currently we have to detect if it is a blocking thread is to check the thread name, which is quite brittle.

Also your put case, I don't see how the read and write would need concurrent blocking threads. The read would be done before then the write would be done afterwards, synchronously.

Reply:
> Just an FYI but the only way currently we have to detect if it is a blocking thread is to check the thread name, which is quite brittle.

Can't we do `!(Thread.currentThread() instanceof ISPNNonBlockingThread)`?

> Also your put case, I don't see how the read and write would need concurrent blocking threads. The read would be done before then the write would be done afterwards, synchronously.

If the cluster executor task does a blocking `cache.put(k, v)`, it needs a (blocking) thread for the entire duration of the cache operation. The read would run on the same thread, but then `continueOnCPUExecutor()` would submit a task to the non-blocking executor, and the next `PersistenceManagerImpl` call would submit a task to the blocking executor.

Reply:
>Can't we do !(Thread.currentThread() instanceof ISPNNonBlockingThread)? 

No, unfortunately. This would include user threads, jgroups etc.

> If the cluster executor task does a blocking cache.put(k, v), it needs a (blocking) thread for the entire duration of the cache operation. The read would run on the same thread, but then continueOnCPUExecutor() would submit a task to the non-blocking executor, and the next PersistenceManagerImpl call would submit a task to the blocking executor.

Oh, okay you were not referring to the read then write. I agree if a blocking operation is invoked on a blocking thread then yes it would use more than 1.